extras/kms/repository_data_key_version.go

```diff
@@ -186,3 +187,129 @@ func (r *repository) ListDataKeyVersions(ctx context.Context, rkvWrapper wrappin
+	if err != nil {
+		return fmt.Errorf("%s: unable to list the current data keys: %w", op, err)
+	}
+	for _, dk := range dks {
+		var versions []*dataKeyVersion
+		if err := r.list(ctx, &versions, "data_key_id = ?", []interface{}{dk.PrivateId}, withReader(reader)); err != nil {
+			return fmt.Errorf("%s: unable to list the current data key versions: %w", op, err)
+		}
+		for _, v := range versions {
+			if err := v.Decrypt(ctx, rkvWrapper); err != nil {
+				return fmt.Errorf("%s: failed to decrypt data key version: %w", op, err)
+			}
+			if err := v.Encrypt(ctx, rkvWrapper); err != nil {
+				return fmt.Errorf("%s: failed to rewrap data key version: %w", op, err)
+			}
+			rowsAffected, err := writer.Update(ctx, v, []string{keyFieldName}, nil, dbw.WithVersion(&v.Version))
+			if err != nil {
+				return fmt.Errorf("%s: failed to update data key version: %w", op, err)
+			}
+			if rowsAffected != 1 {
+				return fmt.Errorf("%s: expected to update 1 data key version and updated %d", op, rowsAffected)
+			}
+		}
+	}
+	return nil
+}
+
+// rotateDataKeyVersionTx will rotate the key version for the given rootKeyId.
+// This function encapsulates all the work required within a dbw.TxHandler and
+// allows this capability to be shared with other repositories or just called
+// within a transaction.  To be clear, this repository function doesn't include
+// its own transaction and is intended to be used within a transaction provide
+// by the caller.
+// Supported options: withRandomReader
+func rotateDataKeyVersionTx(ctx context.Context, reader dbw.Reader, writer dbw.Writer, rootKeyVersionId string, rkvWrapper wrapping.Wrapper, rootKeyId string, purpose KeyPurpose, opt ...Option) error {
+	const op = "kms.rotateDataKeyVersionTx"
+	if isNil(reader) {
+		return fmt.Errorf("%s: missing reader: %w", op, ErrInvalidParameter)
+	}
+	if isNil(writer) {
+		return fmt.Errorf("%s: missing writer: %w", op, ErrInvalidParameter)
+	}
+	if rootKeyVersionId == "" {
+		return fmt.Errorf("%s: missing root key version id: %w", op, ErrInvalidParameter)
+	}
+	if isNil(rkvWrapper) {
+		return fmt.Errorf("%s: missing root key version wrapper: %w", op, ErrInvalidParameter)
+	}
+	if rootKeyId == "" {
+		return fmt.Errorf("%s: missing root key id: %w", op, ErrInvalidParameter)
+	}
+	if purpose == KeyPurposeUnknown {
+		return fmt.Errorf("%s: missing key purpose: %w", op, ErrInvalidParameter)
+	}
+
+	opts := getOpts(opt...)
+	if isNil(opts.withRandomReader) {
+		opts.withRandomReader = rand.Reader
+	}
+
+	r, err := newRepository(reader, writer)
+	if err != nil {
+		return fmt.Errorf("%s: unable to create repo: %w", op, err)
+	}
+	dataKeys, err := r.ListDataKeys(ctx, withPurpose(purpose), withRootKeyId(rootKeyId), withReader(reader))
+	switch {
+	case err != nil:
+		return fmt.Errorf("%s: unable to lookup data key for %q: %w", op, purpose, err)
+	case len(dataKeys) == 0:
+		// this is NOT an error, there's just not data key to rotate for this purpose.
```

Comment:
Wouldn't it be an error to call this function without having an actual data key? It seems like if you don't have a data key for a purpose you shouldn't be calling the rotation function for the data key versions with that purpose.

Reply:
maybe... I was thinking of the situation where you decide to rotate keys before reconciling them.

Reply:
maybe... I was thinking about the scenario where the caller rotates the keys before reconciling them.